plugins/rts54hub/fu-rts54hub-rtd21xx-device.c

```diff
@@ -0,0 +1,618 @@
+		if (tmp <= 0xff) {
+			self->target_addr = tmp;
+			return TRUE;
+		}
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "invalid target address");
+		return FALSE;
+	}
+
+	/* load i2c speed from quirks */
+	if (g_strcmp0 (key, "Rts54I2cSpeed") == 0) {
+		guint64 tmp = fu_common_strtoull (value);
+		if (tmp < FU_RTS54HUB_I2C_SPEED_LAST) {
+			self->i2c_speed = tmp;
+			return TRUE;
+		}
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "invalid I²C speed");
+		return FALSE;
+	}
+
+	/* load register address length from quirks */
+	if (g_strcmp0 (key, "Rts54RegisterAddrLen") == 0) {
+		guint64 tmp = fu_common_strtoull (value);
+		if (tmp <= 0xff) {
+			self->register_addr_len = tmp;
+			return TRUE;
+		}
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "invalid register address length");
+		return FALSE;
+	}
+
+	/* failed */
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_SUPPORTED,
+			     "quirk key not supported");
+	return FALSE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_i2c_write (FuRts54hubRtd21xxDevice *self,
+				      guint8 target_addr, guint8 sub_addr,
+				      const guint8 *data, gsize datasz,
+				      GError **error)
+{
+	FuRts54HubDevice *parent;
+
+	parent = fu_rts54hub_rtd21xx_device_get_parent (self, error);
+	if (parent == NULL)
+		return FALSE;
+	if (!fu_rts54hub_device_vendor_cmd (parent, 1, error))
+		return FALSE;
+
+	if (target_addr != self->target_addr) {
+		if (!fu_rts54hub_device_i2c_config (parent, target_addr, 1,
+						    FU_RTS54HUB_I2C_SPEED_200K,
+						    error))
+			return FALSE;
+		self->target_addr = target_addr;
+	}
+	if (!fu_rts54hub_device_i2c_write (parent, sub_addr, data, datasz, error)) {
+		g_prefix_error (error,
+				"failed to write I2C @0x%02x:%02x: ",
+				target_addr, sub_addr);
+		return FALSE;
+	}
+	g_usleep (I2C_DELAY_AFTER_SEND);
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_i2c_read (FuRts54hubRtd21xxDevice *self,
+				     guint8 target_addr, guint8 sub_addr,
+				     guint8 *data, gsize datasz,
+				     GError **error)
+{
+	FuRts54HubDevice *parent;
+
+	parent = fu_rts54hub_rtd21xx_device_get_parent (self, error);
+	if (parent == NULL)
+		return FALSE;
+	if (!fu_rts54hub_device_vendor_cmd (parent, 1, error))
+		return FALSE;
+	if (target_addr != self->target_addr) {
+		if (!fu_rts54hub_device_i2c_config (parent, target_addr, 1,
+						    FU_RTS54HUB_I2C_SPEED_200K,
+						    error))
+			return FALSE;
+		self->target_addr = target_addr;
+	}
+	if (!fu_rts54hub_device_i2c_read (parent, sub_addr, data, datasz, error)) {
+		g_prefix_error (error, "failed to read I2C: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_read_status_raw (FuRts54hubRtd21xxDevice *self,
+					    guint8 *status,
+					    GError **error)
+{
+	guint8 buf[] = { 0x00 };
```

Comment:
this seems like a bug.  you shouldn't declare the extra variable, but should be using *status in the `*_i2c_read` function.

Reply:
But status is optional...

Reply:
Oh I see.  Can you change the funny `guint8 buf[]` instead to `guint8 buf`?

Reply:
Indeed I can! Fixed.